### PR TITLE
[BEAM-646] Move TransformHierarchy Maintenance into it

### PR DESCRIPTION
**Pull request by tgroh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

This reduces the complexity of Pipeline.applyInternal by keeping the
responsiblities to passing a node into the Transform Hierarchy,
enforcing name uniqueness, and causing the runner to expand the
PTransform. This logic is moved to the appropriate application sites.

### CODE REVIEW

**Comment by tgroh** [comment]
R: @kennknowles

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformTreeNode.java [hunk 1]
```diff
@@ -51,39 +52,49 @@
...
 
   // TODO: track which outputs need to be exported to parent.
   // Output of the transform, in unexpanded form.
   private POutput output;
 
-  private boolean finishedSpecifying = false;
+  @VisibleForTesting
+  boolean finishedSpecifying = false;
+
+  public static TransformTreeNode root(TransformHierarchy hierarchy) {
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformTreeNode.java]: Javadoc

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformTreeNode.java [hunk 2]
```diff
@@ -51,39 +52,49 @@
...
+    return new TransformTreeNode(hierarchy, null, null, "", null);
+  }
+
+  public static TransformTreeNode subtransform(
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformTreeNode.java]: Javadoc

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformTreeNode.java [hunk 3]
```diff
@@ -180,17 +202,10 @@ public POutput getOutput() {
...
-  /**
-   * Returns the transform outputs, in expanded form.
-   */
-  public Collection<? extends PValue> getExpandedOutputs() {
-    if (output != null) {
-      return output.expand();
-    } else {
-      return Collections.emptyList();
-    }
+  AppliedPTransform<?, ?, ?> toAppliedPTransform() {
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformTreeNode.java]: Is this "mostly" deprecated? Or even perhaps fully `@Deprecated`?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformHierarchy.java [hunk 4]
```diff
@@ -34,70 +37,95 @@
...
+    current = node;
+    return current;
   }
 
-  /**
-   * Add a TransformTreeNode to the stack.
-   */
-  public void pushNode(TransformTreeNode current) {
-    transformStack.push(current);
+  public void finishSpecifyingInput() {
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformHierarchy.java]: Javadoc, especially for spec.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformHierarchy.java [hunk 5]
```diff
@@ -34,70 +37,95 @@
...
+    }
   }
 
-  /**
-   * Removes the last TransformTreeNode from the stack.
-   */
-  public void popNode() {
-    transformStack.pop();
-    checkState(!transformStack.isEmpty());
+  public void setOutput(POutput output) {
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformHierarchy.java]: javadoc

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformTreeNode.java (hunk 1 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformTreeNode.java]: Done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformTreeNode.java (hunk 2 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformTreeNode.java]: Done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformTreeNode.java (hunk 3 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformTreeNode.java]: This is linked to `recordAsOutput` + `getProducingTransformInternal` and probably will be removed with those.

Given that it's package private, I'm happy to, at least for now, leave it as an implementation detail of `TransformHeirarchy`

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformHierarchy.java (hunk 4 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformHierarchy.java]: Done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformHierarchy.java (hunk 5 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformHierarchy.java]: Done.

**Pull request closed without merging**